File: Data Structures & Algorithms/decode-ways/submission-4.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = [1] * (len(s) + 1)

        for i in range(len(s) - 1, -1, -1):
            if s[i] == "0":
                dp[i] = 0
            else:
                dp[i] = dp[i + 1]
            
            if i + 1 < len(s) and (s[i] == "1" or s[i] == "2" and s[i + 1] in "0123456"):
                dp[i] += dp[i + 2]
        return dp[0]

        # n = len(s)
        # mem = [-1] * len(s)

        # def dfs(i):
        #     if i == len(s):
        #         return 1
        #     if s[i] == "0":
        #         return 0
            
        #     if mem[i] == -1:
        #         if i + 2 <= n and int(s[i:i + 2]) <= 26:
        #             mem[i] = dfs(i + 2) + dfs(i + 1)
        #         else:
        #             mem[i] = dfs(i + 1)
        #     return mem[i]
            

        # return dfs(0) 
        

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.cpp
class Solution {
public:
    // unordered_map<string, int> memo;

    // int lengthOfLIS(vector<int>& nums) {
    //     return dfs(nums, 0, -1);
    // }

    // int dfs(vector<int>& nums, int i, int curr) {
    //     if (i >= nums.size()) {
    //         return 0;
    //     }
    //     if (memo.find(to_string(i) + "," + to_string(curr)) != memo.end()) {
    //         return memo[to_string(i) + "," + to_string(curr)];
    //     }
    //     if (curr == -1 || nums[i] > nums[curr]) {
    //         memo[to_string(i) + "," + to_string(curr)] =
    //          max(dfs(nums, i + 1, curr), 1 + dfs(nums, i + 1, i));
    //     }
    //     else {
    //         memo[to_string(i) + "," + to_string(curr)] = dfs(nums, i + 1, curr);
    //     }

    //     return memo[to_string(i) + "," + to_string(curr)];
    // }

    int lengthOfLIS(vector<int>& nums) {
        vector<int> LIS(nums.size(), 1);

        for (int i = nums.size() - 1; i >= 0; i--) {
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[i] < nums[j]) {
                    LIS[i] = max(LIS[i], 1 + LIS[j]);
                }
            }
        }

        return *max_element(LIS.begin(), LIS.end());
    }
};
